packages/js/client/src/wasm/WasmWeb3Api.ts

```diff
@@ -0,0 +1,377 @@
+    }
+
+    switch (state) {
+      case "Abort": {
+        return {
+          error: new Error(
+            `WasmWeb3Api: Thread aborted execution.\nMessage: ${abortMessage}`
+          ),
+        };
+      }
+      case "LogQueryError": {
+        return {
+          error: new Error(
+            `WasmWeb3Api: invocation exception encourtered.\n` +
+              `uri: ${this._uri.uri}\nmodule: ${module}\n` +
+              `method: ${method}\n` +
+              `input: ${JSON.stringify(input, null, 2)}\n` +
+              `exception: ${queryError}`
+          ),
+        };
+      }
+      case "LogQueryResult": {
+        if (decode) {
+          return { data: MsgPack.decode(queryResult as ArrayBuffer) };
+        } else {
+          return { data: queryResult };
+        }
+      }
+      default: {
+        throw Error(`WasmWeb3Api: Unknown state "${state}"`);
+      }
+    }
+  }
+
+  public async getSchema(client: Client): Promise<string> {
+    if (this._schema) {
+      return this._schema;
+    }
+
+    const module = this._manifest.query || this._manifest.mutation;
+
+    if (!module) {
+      // TODO: this won't work for abstract APIs
+      throw Error(`WasmWeb3Api: No module was found.`);
+    }
+
+    const { data, error } = await ApiResolver.Query.getFile(
+      client,
+      this._apiResolver,
+      path.join(this._uri.path, module.schema.file)
+    );
+
+    if (error) {
+      throw error;
+    }
+
+    // If nothing is returned, the schema was not found
+    if (!data) {
+      throw Error(
+        `WasmWeb3Api: Schema was not found.\nURI: ${this._uri}\nSubpath: ${module.schema.file}`
+      );
+    }
+
+    const decoder = new TextDecoder();
+    this._schema = decoder.decode(data);
+
+    if (!this._schema) {
+      throw Error(
+        `WasmWeb3Api: Decoding the schema's bytes array failed.\nBytes: ${data}`
+      );
+    }
+
+    return this._schema;
+  }
+
+  private getTypeInfo(schema: string): TypeInfo {
+    if (!this._typeInfo) {
+      this._typeInfo = parseSchema(schema);
+    }
+
+    return this._typeInfo;
+  }
+
+  private async getWasmModule(
+    module: InvokableModules,
+    client: Client
+  ): Promise<ArrayBuffer> {
+    if (this._wasm[module] !== undefined) {
+      return this._wasm[module] as ArrayBuffer;
+    }
+
+    const moduleManifest = this._manifest[module];
+
+    if (!moduleManifest) {
+      throw Error(
+        `Package manifest does not contain a definition for module "${module}"`
+      );
+    }
+
+    const { data, error } = await ApiResolver.Query.getFile(
+      client,
+      this._apiResolver,
+      path.join(this._uri.path, moduleManifest.module.file)
+    );
+
+    if (error) {
+      throw error;
+    }
+
+    // If nothing is returned, the module was not found
+    if (!data) {
+      throw Error(
+        `Module was not found.\nURI: ${this._uri}\nSubpath: ${moduleManifest.module.file}`
+      );
+    }
+
+    this._wasm[module] = data;
+    return data;
+  }
+}
```

Comment:
not sure what's going on here or why there is manual data transfer but I found working with treads is much simpler with https://threads.js.org/

Reply:
The first implementation was actually using Threads.js, but I had to switch since it doesn't support passing SharedArrayBuffer as a worker data argument. Here's the commit where I made the transition: https://github.com/Web3-API/prototype/pull/105/commits/855ea50f83428a910411685401d2c45637a63b20

Reply:
What is going on here is:
- Create a thread for the WASM module we're about to call into
- The thread can dispatch actions for the main thread to do
- When the thread dispatches a "sub-invoke" action, it immediately pauses itself to stop the WASM VM from continuing execution. The main thread then performs the sub-invocation, passes the data back into the thread using a SharedArrayBuffer, and the thread continues its execution given its sub-invoke result.

Reply:
Tnx for explanation. But isn't `Transfer(ArrayBuffer)` basically SharedArrayBuffer: https://threads.js.org/usage-advanced#transferable-objects ?

Reply:
No this will not work since you need to have a valid reference to the ArrayBuffer on both ends (main-thread + wasm-thread). This is required so that you can use `Atomics.wait(...)` within the wasm-thread, and `Atomics.notify(...)` in the main-thread. Both of these require a reference to a SharedArrayBuffer.